### PR TITLE
Fix bugs and add new features

### PR DESCRIPTION
**Pull request by ImpSy**

### CODE REVIEW

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 1]
```diff
@@ -77,11 +81,16 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
   /**
     * Send a POST request to Data Mechanics collector API ("the server")
     *
-    * - Anything other than a 200 status code counts as failed.
     * - Handles access token
+    * - Status Code:
+    *   -> 200: Request is a success
+    *   -> 429: RateLimit as been reach for this app
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: s/as been reach for/has been reached for

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 2]
```diff
@@ -97,15 +106,18 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
...
-      var errorMessage = s"Status $statusCode: ${httpResponse.getStatusLine.getReasonPhrase}."
-      apiReturnMessage.foreach(
-        m => errorMessage += s" ${m}."
-      )
-      throw new IOException(errorMessage)
+    statusCode match {
+      case 200 => logInfo(successMsg)
+      case 429 =>
+        isRateLimited.set(true)
+        logError(s"RateLimit has been reach, collection will now stop")
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: reached

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 3]
```diff
@@ -97,15 +106,18 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
...
+      case _ =>
+        var errorMessage = s"Status $statusCode: ${httpResponse.getStatusLine.getReasonPhrase}."
+        apiReturnMessage.foreach(
+          m => errorMessage += s" ${m}."
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: remove curly brackets

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 4]
```diff
@@ -118,8 +130,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
   def sendHeartbeat(): Unit = {
     val url = s"$collectorURL/heartbeat"
     try {
-      sendRequest(httpClientHeartbeat, url, dmAppId.toJson)
-      logInfo(s"Successfully sent heartbeat")
+      sendRequest(httpClientHeartbeat, url, dmAppId.toJson, s"Successfully sent heartbeat")
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: Remove `s` (string interpolation is not needed here)

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 5]
```diff
@@ -136,8 +147,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
     val url = s"$collectorURL/ack"
 
     try {
-      sendRequest(httpClient, url, dmAppId.toJson)
-      logInfo(s"Successfully sent ack")
+      sendRequest(httpClient, url, dmAppId.toJson, s"Successfully sent ack")
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: Remove `s`

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 6]
```diff
@@ -155,8 +165,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
     val url = s"$collectorURL/bulk"
 
     try {
-      sendRequest(httpClient, url, payload.toJson)
-      logInfo(s"Successfully sent payload")
+      sendRequest(httpClient, url, payload.toJson, s"Successfully sent payload")
```
Comment (jaceklaskowski) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: nit: same

**Review by jaceklaskowski — APPROVED** [review]
LGTM

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 7]
```diff
@@ -237,6 +238,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
       logWarning(s"Stopped waiting for pending events to be sent (max wait duration is ${maxWaitOnEnd}), although $nbPendingEvents were remaining")
     }
     sendAck()
+    logInfo(s"Application will be available in a few minutes on Delight at this url: $delightURL/apps/$dmAppId")
```
Comment (jrj-d) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala:241]:
```
logInfo(s"Application will be available a few minutes after it completes on Delight at this url: $delightURL/apps/$dmAppId")
```

Comment (jrj-d) [reply src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala:241]: Sorry I thought this was at the beginning of the streaming, while it is at the end. Please discard!

File: src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 8]
```diff
@@ -308,6 +310,7 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
       }
       heartbeatThread.start()
       logInfo("Started DelightStreamingConnector heartbeat thread")
+      logInfo(s"Application will be available once your app is terminated on Delight at this url: $delightURL/apps/$dmAppId")
```
Comment (jrj-d) [inline src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: logInfo(s"Application will be available on Delight a few minutes after it completes at this url: $delightURL/apps/$dmAppId")

**Review by jrj-d — APPROVED** [review]
Wording nit. Otherwise it looks good to me!

Thanks @jaceklaskowski for the review :)

**Pull request merged**